system/security/shared/secmanagertracedecorator.hpp

```diff
@@ -0,0 +1,364 @@
+ * interface. In this case, decorated_t should be CLdapSecManager and decorated_interface_t should
+ * be ISecManager.
+ */
+template <typename decorated_t, typename decorated_interface_t = decorated_t>
+class TDecorator : public CInterfaceOf<decorated_interface_t>
+{
+protected:
+    Linked<decorated_t> decorated;
+public:
+    TDecorator(decorated_t& _decorated) : decorated(&_decorated) {}
+    virtual ~TDecorator() {}
+    decorated_t* queryDecorated() { return decorated.get(); }
+    decorated_t* getDecorated() { return decorated.getLink(); }
+};
+
+/**
+ * @brief Macro used start tracing a block of code in the security manager decorator.
+ *
+ * Create a new named internal span and enter a try block. Used with END_SEC_MANAGER_TRACE_BLOCK,
+ * provides consistent timing and exception handling for the inned code block.
+ *
+ * Some security manager requests include outgoing remote calls, but the ESP does not assume
+ * which requests are remote. Managers making remote calls may considier creating client spans
+ * as needed.
+ */
+#define START_SEC_MANAGER_TRACE_BLOCK(name) \
+    OwnedSpanScope spanScope(queryThreadedActiveSpan()->createInternalSpan(name)); \
+    try \
+    {
+
+/**
+ * @brief Macro used to end tracing a block of code in the security manager decorator.
+ *
+ * Ends a try block and defines standard exception handling. Use with START_SEC_MANAGER_TRACE_BLOCK,
+ * provides consistent timing and exception handling for the timed code block.
+ */
+#define END_SEC_MANAGER_TRACE_BLOCK \
+    } \
+    catch (IException* e) \
+    { \
+        spanScope->recordException(e); \
+        throw; \
+    } \
+    catch (...) \
+    { \
+        Owned<IException> e(makeStringException(-1, "unknown exception")); \
+        spanScope->recordException(e); \
+        throw; \
+    }
+
+/**
+ * @brief Decorator for ISecManager that adds tracing to the interface.
+ *
+ * Tracing is added to selected methods of ISecManager. Traced methods create a span for the
+ * duration of the decorated method. Untraced methods are passed through to the decorated object.
+ *
+ * The default decorated type is sufficient for most security managers. CLdapSecManager is an
+ * exception because platform processes depend on interfaces not included in the default.
+ * - If ISecManager and CLdapSecManager interfaces are standardized, this point becomes moot.
+ * - If ISecManager is extended to declare LDAP-specific interfaces, and CLdapSecManager implements
+ *   this interface, create a subclass of TSecManagerTraceDecorator to decorate the new interface
+ *   and replace ISecManager with the new interface name as template parameters.
+ * - If nothing else changes, create a subclasses of TSecManagerTraceDecorator, with template
+ *   parameters CLdapSecManager and ISecManager, to decorate the LDAP-specific interfaces.
+ */
+template <typename secmgr_t = ISecManager, typename secmgr_interface_t = ISecManager>
```

Comment:
what would be the advantage/disadvantage of this approach vs a base implementation which all sec managers could extend?

Reply:
Without the template, the base class holds a reference to an ISecManager instance, and any subclass holds a second reference to the same object but typed as the manager implementation class. Very minor, but I'm not a fan of holding the same value twice. The first template parameter allows the base class to hold the type needed by the subclass.

The two template parameters could be consolidated into one, reducing the complexity, if either:
- The ISecManager and CLdapSecManager interfaces are standardized.
- A new extension of ISecManager declaring the LDAP extensions is used.

I'm not married to either approach.